Policy employees merge refactor

chiragsalian opened the pull request:
Held on https://github.com/Expensify/Web-Expensify/pull/34602

<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/Expensify/issues/211556

### Tests
# Admin test
1. Log into the same admin account in two browsers. One normal browser and another incognito.
2. On both browsers. Click on Avatar icon -> Select any workspace -> Manage members.
3. On one browser add a member and confirm it was successfully added in the normal browser and that it shows up immediately in the incognito browser indicating that the push notification worked well.

https://user-images.githubusercontent.com/14356145/190027012-b5337403-c3a6-4248-98cf-6a7f64f4cb81.mov

## Admin and new user test.
1. Repeat the same steps above with the only difference being the admin logs into normal browser while the new user logs into incognito browser.
2. Confirm the new user immediately gets the new workspace chat and sees contact information of the other members on the same policy

https://user-images.githubusercontent.com/14356145/190027242-dcbacdd9-f879-4832-b544-c7cfa91d9455.mov

## Admin and existing user test
1. Repeat the same steps above with the only difference being the admin logs into normal browser while the existing user logs into incognito browser.
2. When a new user is added confirm the existing user can immediately find the new user in new chat easily using first and last name.

https://user-images.githubusercontent.com/14356145/190027413-9e9862c5-c6ba-44df-8c13-21739895de1f.mov

- [x] Verify that no errors appear in the JS console
I see a JS console error but it's not related to my changes.

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Same as Tests.

- [ ] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

chiragsalian committed:
initial code

chiragsalian committed:
Merge branch 'main' into chirag-Policy_Employees_Merge-refactor

chiragsalian committed:
testing addMembersToWorkspace

chiragsalian committed:
Merge branch 'main' into chirag-Policy_Employees_Merge-refactor

chiragsalian committed:
cleanup and navigate back

chiragsalian committed:
updating WorkspaceMembersPage to display member list

chiragsalian committed:
Merge branch 'main' into chirag-Policy_Employees_Merge-refactor

chiragsalian committed:
using key email to future proof code so role can be passed easily in future

chiragsalian committed:
removing unnecessary params

chiragsalian committed:
Merge branch 'main' into chirag-Policy_Employees_Merge-refactor

chiragsalian committed:
Merge branch 'main' into chirag-Policy_Employees_Merge-refactor

chiragsalian committed:
Adding translations

iwiznia commented on src/libs/actions/Policy.js:
```diff
@@ -244,51 +243,57 @@ function removeMembers(members, policyID) {
-    Onyx.merge(`${ONYXKEYS.COLLECTION.POLICY_MEMBER_LIST}${policyID}`, employeeUpdate);
+    const optimisticData = [
+        {
+            onyxMethod: 'merge',
```
Use the constant (same in the other 3) https://github.com/Expensify/App/blob/23f4fc80ca4a2df155a58dfc51eef301125a1fc3/src/CONST.js#L794

chiragsalian replied:
whoops yes

iwiznia commented on src/libs/actions/Policy.js:
```diff
@@ -244,51 +243,57 @@ function removeMembers(members, policyID) {
-    // Make the API call to merge the login into the policy
-    DeprecatedAPI.Policy_Employees_Merge({
+            // Convert to object with each key containing {pendingAction: ‘add’}
+            value: _.object(logins, Array(logins.length).fill({pendingAction: 'add'})),
```
Use the constant https://github.com/Expensify/App/blob/23f4fc80ca4a2df155a58dfc51eef301125a1fc3/src/CONST.js#L800

chiragsalian committed:
Merge branch 'main' into chirag-Policy_Employees_Merge-refactor

chiragsalian committed:
Update src/languages/es.js

Co-authored-by: Ionatan Wiznia <ionatan@expensify.com>

chiragsalian committed:
using const

chiragsalian committed:
Merge branch 'chirag-Policy_Employees_Merge-refactor' of github.com:Expensify/App into chirag-Policy_Employees_Merge-refactor

chiragsalian commented:
Ready for re-review

MonilBhavsar reviewed (commented):
Looks good! Will test once the Web-E PR discussions are resolved and merged

chiragsalian committed:
Merge branch 'main' into chirag-Policy_Employees_Merge-refactor

MonilBhavsar reviewed (commented):
Tested with Web-PR.
The tests mentioned in the tests section works well, but doesn't work offline. @chiragsalian it is should be pattern B, right? Could you please check If I'm holding something wrong or we need some more changes here.

chiragsalian committed:
enabledWhenOffline prop

chiragsalian committed:
showing opacity for add

chiragsalian commented:
> The tests mentioned in the tests section works well, but doesn't work offline.  @chiragsalian it is should be pattern B, right?

Yup its pattern B, i was just hoping if the data was correct it would just work. I retested and there are few things missing,
1. The invite button is blocked when offline - I added a prop called enabledWhenOffline to address this.
2. The pendingAction for "add" was not working to add opacity so I fixed that.
3. Finally the data provided to WorkspaceMembersPage does not contain all the necessary information and that I realized is done [here](https://github.com/Expensify/App/pull/10452/files#diff-bbfbd221fc2e98be145adb346139ceb43a56501e1feef799db83b51ae672421eR286-R299) in ioni's PR so you can can copy over the same to test pendingActions. I'm not copying it over as its dupe work.

So it should be good now. Let me know if those make sense. I was able to test the PR offline.

chiragsalian commented on src/components/OfflineWithFeedback.js:
```diff
@@ -74,10 +74,7 @@ function applyStrikeThrough(children) {
-    const isUpdateOrDeleteError = hasErrors && (props.pendingAction === 'delete' || props.pendingAction === 'update');
-    const isAddError = hasErrors && props.pendingAction === 'add';
-    const needsOpacity = (isOfflinePendingAction && !isUpdateOrDeleteError) || isAddError;
+    const needsOpacity = props.network.isOffline && props.pendingAction === 'add';
```
@iwiznia, can you double check this code. As per the [pattern B](https://docs.google.com/document/d/1Ija9CHiNBH1HVnnITA-XuohF31BoiseY18a8_BxFtjc/edit#heading=h.mnjguil21u0a) doc pendingAction 'add' should show up with opacity when offline. So I simplified the logic here.

The previous logic here shows that if pendingAction is delete or update with errors then don't show opacity. The updated logic is pretty much the same where nothing happens wrt to opacity for delete or update, but will show opacity for add if its offline if its an error or not.

Let me know if this is alright or if I'm missing something.

iwiznia replied:
So, something was broken? If not, I would say leave it as it was.

> The previous logic here shows that if pendingAction is delete or update with errors then don't show opacity.

This is correct, because if update or delete failed, it means the item does exist in the DB, so no opacity is needed.

MonilBhavsar commented on src/components/OfflineWithFeedback.js:
```diff
@@ -74,10 +74,7 @@ function applyStrikeThrough(children) {
-    const isUpdateOrDeleteError = hasErrors && (props.pendingAction === 'delete' || props.pendingAction === 'update');
-    const isAddError = hasErrors && props.pendingAction === 'add';
-    const needsOpacity = (isOfflinePendingAction && !isUpdateOrDeleteError) || isAddError;
+    const needsOpacity = props.network.isOffline && props.pendingAction === 'add';
```
`needsOpacity` condition should include `hasErrors`, no?

MonilBhavsar replied:
Also, I can't see opacity when trying to add member offline. I seems tricky to debug as browser's react debugger is not updating when I'm offline and when I turn online there is no pendingAction 😅

chiragsalian replied:
> needsOpacity condition should include hasErrors, no?

I'm unsure but looking at the previous logic I would say no because there are three types or errors - add, delete & update. For add, regardless if its error or not I'm showing opacity. In the previous code,
> const needsOpacity = (isOfflinePendingAction && !isUpdateOrDeleteError) || isAddError;

needsOpacity is false for update or delete errors when offline. So I removed that since it does show opacity in that flow anyway. But either way I've asked for logic confirmation to ioni above so maybe its easier to keep that discussion in a single thread above?

chiragsalian replied:
updated to revert the logic cause after discussing with ioni I realized it just works when adding anyway.

MonilBhavsar commented:
> Finally the data provided to WorkspaceMembersPage does not contain all the necessary information and that I realized is done [here](https://github.com/Expensify/App/pull/10452/files#diff-bbfbd221fc2e98be145adb346139ceb43a56501e1feef799db83b51ae672421eR286-R299) in ioni's PR so you can can copy over the same to test pendingActions. I'm not copying it over as its dupe work.

Sorry, I didn't get this. Are both PR's interdependent? 
Also, can we please include some tests/QA for offline mode too 🙏

JmillsExpensify commented:
Web-E 34602 was merged and is on staging, so should hit production on Monday. I removed the hold but be careful until then.

chiragsalian commented:
> Sorry, I didn't get this. Are both PR's interdependent?

For online testing they are not dependent on each other. For offline testing, yes there is a block of code [here](https://github.com/Expensify/App/pull/10452/files#diff-bbfbd221fc2e98be145adb346139ceb43a56501e1feef799db83b51ae672421eR286-R299) ioni uses which the PR here is dependent off. I've explain my reason why I've not copied it over here,
> Finally the data provided to WorkspaceMembersPage does not contain all the necessary information and that I realized is done [here](https://github.com/Expensify/App/pull/10452/files#diff-bbfbd221fc2e98be145adb346139ceb43a56501e1feef799db83b51ae672421eR286-R299) in ioni's PR so you can can copy over the same to test pendingActions. I'm not copying it over as its dupe work.

> Also, can we please include some tests/QA for offline mode too 🙏

The reason I've not included it is because its dependent on ioni's PR. We normally just write the normal test case ([example](https://github.com/Expensify/App/pull/10452)) for QA. I could copy over ioni's code and write up offline steps but then it sorta feels like dupe work which will cause an unnecessary code conflict. Anyway I don't feel too strongly on copying the code but let me know if you feel strongly about it.

iwiznia commented on src/libs/actions/Policy.js:
```diff
@@ -244,51 +243,57 @@ function removeMembers(members, policyID) {
+
+            // Convert to object with each key clearing pendingAction. We don’t
+            // need to remove the members since that will be handled by onClose of OfflineWithFeedback.
+            value: _.object(logins, Array(logins.length).fill({pendingAction: null})),
```
You need to remove the errors here, no?

chiragsalian replied:
Not really, how can a login here have error? Are you considering the case something got an error and was added again and hence we should remove the error?

iwiznia replied:
> Not really, how can a login here have error?

You tried something once, it did not work (so there's an error), you try again and it works.

chiragsalian replied:
yup, i think you are correct. Updating.

chiragsalian committed:
code revert

chiragsalian committed:
clearing errors on success

chiragsalian commented:
Updated and ready for review.

iwiznia commented on src/components/FormAlertWithSubmitButton.js:
```diff
@@ -51,12 +55,14 @@ const FormAlertWithSubmitButton = props => (
         onFixTheErrorsPressed={props.onFixTheErrorsPressed}
     >
-        {isOffline => (isOffline ? (
+        {isOffline => ((isOffline && !props.enabledWhenOffline) ? (
```
I don't get this... both sides of the ternary are rendering the exact same component now.

chiragsalian replied:
🤦‍♂️, oops i didn't correct the logic after testing it out. I'm pushing the correct logic. which is basically reverting the button props and I tested it out.

chiragsalian replied:
Updated and retested

chiragsalian committed:
logic fixes

chiragsalian commented:
Updated

MonilBhavsar reviewed (approved):
Looks good and works well! Please add QA steps 🙏

chiragsalian commented:
Yup thank you, i added to mention its the same as tests.

chiragsalian commented:
Not emergency, tests had passed just fine.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3153866255) to staging by @chiragsalian in version: 1.2.10-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3153866255) to staging by @chiragsalian in version: 1.2.10-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3161850641) to production by @roryabraham in version: 1.2.10-1 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
